### PR TITLE
#870 stocktake reasons 01

### PR DESCRIPTION
**Pull request by roxy-dao**
Part of #870 
- Added `inventory_adjustment_reason` table
- `inventory_adjustment_reason_id` fields on `stocktake_line` and `invoice_line`
- Sync `inventory_adjustment_reason` table to mSupply `options` table 
- Named type **positive** and **negative** to go with mSupply, but can change to **reduction/addition** if this is preferred 
- Reasons are currently mandatory if active and there is a function to check if the reason is valid
- Tests for stocktake reasons

### CODE REVIEW

File: server/repository/src/migrations/v1_01_01/mod.rs [hunk 1]
```diff
@@ -29,6 +29,43 @@ impl Migration for V1_01_01 {
                 WHERE name_store_join.name_id IN (SELECT name_id FROM store WHERE store.id = name_store_join.store_id);"#
         )?;
 
+        #[cfg(not(feature = "postgres"))]
```
Comment (roxy-dao) [inline server/repository/src/migrations/v1_01_01/mod.rs:32]: Will move migrations to v01_01_02 once #1044 is merged

File: server/service/src/sync/sync_buffer.rs [hunk 2]
```diff
@@ -41,7 +42,7 @@ impl<'a> SyncBuffer<'a> {
         }
     }
 
-    pub(crate) fn record_successfull_integration(
+    pub(crate) fn record_successful_integration(
```
Comment (roxy-dao) [inline server/service/src/sync/sync_buffer.rs:45]: Off track typo 😄

File: server/service/src/stocktake_line/insert.rs [hunk 3]
```diff
@@ -164,6 +183,23 @@ fn validate(
         }
     }
 
+    let stocktake_difference = stocktake_difference(&input.counted_number_of_packs, &stock_line);
+    if check_active_adjustment_reasons(connection, stocktake_difference)?.is_some()
+        && input.inventory_adjustment_reason_id.is_none()
+        && stocktake_difference != 0.0
```
Comment (roxy-dao) [inline server/service/src/stocktake_line/insert.rs:189]: Or 
```suggestion
        && input.counted_number_of_packs.is_none()
```
🤷🏻

File: server/service/src/stocktake_line/insert.rs [hunk 4]
```diff
@@ -112,6 +117,20 @@ fn check_stock_line_exists(
         .pop())
 }
 
+pub fn stocktake_difference(
```
Comment (mark-prins) [inline server/service/src/stocktake_line/insert.rs:120]: for me, the name of the fn doesn't fully describe it - a difference doesn't describe the change, just that something is different. more helpful to have `stocktake_reduction_amount` - that way it is clearer what a positive or negative result would mean. I was unsure when using the fn whether it was referring to stock - counted or the other way around.

File: server/service/src/stocktake_line/insert.rs (hunk 3 above)
Comment (mark-prins) [reply server/service/src/stocktake_line/insert.rs:189]: the two lines aren't equivalent 😕 - I would run with `stocktake_difference != 0.0` that makes sense to me

**Review by mark-prins — APPROVED** [review]
👍  LGTM! just one postgres compilation issue to fix before merging

File: server/repository/src/migrations/v1_01_01/mod.rs [hunk 5]
```diff
@@ -29,6 +29,43 @@ impl Migration for V1_01_01 {
...
+        #[cfg(feature = "postgres")]
+        const INVENTORY_ADJUSTMENT_REASON_TYPE: &'static str = "inventory_adjustment_type";
+        #[cfg(feature = "postgres")]
+        sql!(
+            connection,
+            r#"
+                CREATE TYPE {INVENTORY_ADJUSTMENT_REASON_TYPE} AS ENUM (
+                    'positive',
+                    'negative',
+                );
```
Comment (mark-prins) [inline server/repository/src/migrations/v1_01_01/mod.rs:43]: These need to be UPPERCASE they come through as upper when I try to sync

```suggestion
                CREATE TYPE {INVENTORY_ADJUSTMENT_REASON_TYPE} AS ENUM (
                    'POSITIVE',
                    'NEGATIVE',
                );
```

**Pull request merged**